src/internal/document_api/mod.rs

```diff
@@ -435,53 +508,163 @@ pub fn encrypt_document<
+                    document_create(
+                        &auth,
+                        r,
+                        doc_id,
+                        &document_name,
+                        [key_errs, encryption_errs].concat(),
+                    )
+                })
+        })
+}
+
+type UserMasterPublicKey = PublicKey;
+/// Get the public keys for a document grant.
+///
+/// # Arguments
+/// `auth`          - info to make webservice requests
+/// `user_grants`   - list of user ids to which document access should be granted
+/// `group_grants`  - list of groups ids to which document access should be granted
+/// `policy_grant`  - policy to apply for document access
+/// `maybe_user_master_pub_key`
+///                 - if Some, contains the logged in user's master public key for self-grant
+///
+/// # Returns
+/// A Future that will resolve to:
+/// (Left)  list of keys for all users and groups that should be granted access to the document
+/// (Right) errors for any invalid users/groups that were passed
+fn resolve_keys_for_grants<'a>(
+    auth: &'a RequestAuth,
+    user_grants: &'a Vec<UserId>,
+    group_grants: &'a Vec<GroupId>,
+    policy_grant: Option<&'a PolicyGrant>,
+    maybe_user_master_pub_key: Option<&'a UserMasterPublicKey>,
+) -> impl Future<Item = (Vec<WithKey<UserOrGroup>>, Vec<DocAccessEditErr>), Error = IronOxideErr> + 'a
+{
     internal::user_api::get_user_keys(auth, user_grants)
-        .join4(
+        .join3(
+            // TICKET: only make these calls if the vecs are non-empty
             internal::group_api::get_group_keys(auth, group_grants),
             policy_grant.map(|p| requests::policy_get::policy_get_request(auth, p)),
-            aes::encrypt_future(rng, &plaintext.to_vec(), *doc_sym_key.bytes()),
         )
-        .and_then(move |(users, groups, maybe_policy_res, encrypted_doc)| {
+        .map(move |(users, groups, maybe_policy_res)| {
             let (group_errs, groups_with_key) = process_groups(groups);
             let (user_errs, users_with_key) = process_users(users);
-            let explicit_grants = if grant_to_author {
-                let self_grant = WithKey::new(
-                    UserOrGroup::User {
-                        id: auth.account_id.clone(),
-                    },
-                    user_master_pub_key.clone(),
-                );
-                [&[self_grant], &users_with_key[..], &groups_with_key[..]].concat()
-            } else {
-                [&users_with_key[..], &groups_with_key[..]].concat()
-            };
+            let explicit_grants = [users_with_key, groups_with_key].concat();
             let (policy_errs, applied_policy_grants) = match maybe_policy_res {
                 None => (vec![], vec![]),
                 Some(res) => process_policy(&res),
             };
+            let maybe_self_grant = {
+                if let Some(user_master_pub_key) = maybe_user_master_pub_key {
```

Comment:
To me this makes more sense to do this with a `map_or`, but if that doesn't compute for you I won't be hurt.

Reply:
I can see that, but I think I'll just leave it as is.